src/components/virtualRepeater/virtualRepeater.js

```diff
@@ -0,0 +1,452 @@
+/**
+ * @ngdoc module
+ * @name material.components.virtualRepeat
+ */
+angular.module('material.components.virtualRepeat', [
+  'material.core',
+  'material.core.gestures'
+])
+.directive('mdVirtualRepeatContainer', VirtualRepeatContainerDirective)
+.directive('mdVirtualRepeat', VirtualRepeatDirective);
+
+
+/**
+ * @ngdoc directive
+ * @name mdVirtualRepeatContainer
+ * @module material.components.virtualRepeat
+ * @restrict E
+ * @description
+ * `md-virtual-repeat-container` provides the scroll container for md-virtual-repeat.
+ *
+ * Virtual repeat is a limited substitute for ng-repeat that allows renders only
+ * enough dom nodes to fill the container and recycling them as the user scrolls.
+ *
+ * @usage
+ * <hljs lang="html">
+ * <md-virtual-repeat-container md-horizonal>
+ *   <div md-virtual-repeat="i in items">Hello {{i}}!</div>
+ * </md-virtual-repeat-containr>
+ *
+ * <md-virtual-repeat-container md-horizonal>
+ *   <div md-virtual-repeat="i in items">Hello {{i}}!</div>
+ * </md-virtual-repeat-containr>
+ * </hljs>
+ *
+ * @param {boolean=} md-horizontal Whether the container should scroll horizontally
+ *     (defaults to scrolling vertically).
+ */
+function VirtualRepeatContainerDirective() {
+  return {
+    controller: VirtualRepeatContainerController,
+    replace: true,
+    require: 'virtualRepeatContainer',
+    restrict: 'E',
+    template: VirtualRepeatContainerTemplate
+  };
+}
+
+
+function VirtualRepeatContainerTemplate($element, $attrs) {
+  var innerHtml = $element[0].innerHTML;
+  $element[0].innerHTML = '';
+  // TODO: we probably don't need the container anymore.
+  return '<div class="md-virtual-repeat-container ' +
+      ($attrs.mdHorizontal ? 'md-horizontal' : 'md-vertical') + '">' +
+    '<div class="md-virtual-repeat-scroller">' +
+      '<div class="md-virtual-repeat-sizer"></div>' +
+      '<div class="md-virtual-repeat-offsetter">' +
+        innerHtml +
+      '</div>' +
+    '</div>' +
+  '</div>';
+}
```

Comment:
Like ng-repeat, I want to avoid creating new scopes except for on the repeated items (which themselves should be non-isolate child scopes). I can probably get rid of 'replace', but I pretty much need this in order for a lot of use-cases to work properly.